Implements Request.pause() / resume()

chdh opened the pull request:
This pull request adds the two methods Request.pause() and Request.resume() to the public Tedious API. They can be used to pause and resume the flow of data rows from a query result. A test case is included.

See #181 for the reasons why this extension is important.

Congelli501 commented:
I tried you branch, and I have the following error using this code:

```javascript
const Connection = require('tedious').Connection;
const Request = require('tedious').Request;

const connection = new Connection({
        server: '???',
        userName: '???',
        password: '???',
        options: {
                database: '???'
        }
});

connection.on('connect', (err) => {
        if (err) {
                throw err;
        }

        const request = new Request("select * from RANDOM_TABLE", (err2, rowCount) => {
                if (err2) {
                        throw err2;
                }

                console.log(rowCount + ' rows');
                connection.close();
        });

        request.on('row', (columns) => {
                console.log(columns.length);
                request.pause();
                setTimeout(() => {
                        request.resume();
                }, 500);
        });

        connection.execSql(request);
});
```

Error:
```
events.js:161
      throw er; // Unhandled 'error' event
      ^

Error: Received 'row' when no sqlRequest is in progress
    at Parser.<anonymous> (/root/node_tedious/node_modules/tedious/lib/connection.js:459:32)
    at emitOne (events.js:96:13)
    at Parser.emit (events.js:189:7)
    at Parser.<anonymous> (/root/node_tedious/node_modules/tedious/lib/token/token-stream-parser.js:54:15)
    at emitOne (events.js:96:13)
    at Parser.emit (events.js:189:7)
    at Parser.Readable.read (/root/node_tedious/node_modules/tedious/node_modules/readable-stream/lib/_stream_readable.js:391:26)
    at flow (/root/node_tedious/node_modules/tedious/node_modules/readable-stream/lib/_stream_readable.js:739:34)
    at resume_ (/root/node_tedious/node_modules/tedious/node_modules/readable-stream/lib/_stream_readable.js:722:3)
    at _combinedTickCallback (internal/process/next_tick.js:74:11)
```

The table contains 20 rows. I tried to pause the stream every 5 rows, but I had the same error.

I'm using node 7.6 on Ubuntu x64.
The SQL server has the following version:
```
Microsoft SQL Server 2005 - 9.00.1399.06 (Intel X86) 
        Oct 14 2005 00:33:37 
        Copyright (c) 1988-2005 Microsoft Corporation
        Developer Edition on Windows NT 6.1 (Build 7601: Service Pack 1)
```

Am I misusing the feature ?
Do you want me to open a bug request for that ?

Don't hesitate if you wan any more feedback / test.

chdh commented:
@Congelli501 Thanks for your bug report. I will analyze the problem and provide a fix.

chdh commented:
@Congelli501 I have fixed the problem. It happened when there were only a few records left when pause() was called. I have added another test case to verify that this problem is solved. Could you please test again?

arthurschreiber commented:
Hey @chdh,

Thank you very much for this PR - it definitely shows that pausing and resuming requests is possible  in `tedious` with a bit of tweaking here and there. 👍 

But as you mentioned over in #512:

> I think that the internal architecture of the TDS protocol driver should be reworked.

I think so as well. Tedious internals should be a bunch of streams that are piped into each other, so that `Request#pause()` would be implemented in terms of pausing the "last" stream in this pipe. If I understand the way Node.JS streams work correctly, this should in turn cause all other streams to be paused due to reaching the `highWatermark` and automatic backpressure handling kicking in.

Would you be interested in helping to transform the tedious internals for this? I have played with a few ideas for this, and we could open an issue to discuss this further. If not, I believe these changes could be merged as they are, I'm just concerned that doing so might make tedious internals even more complex (and fragile) than they already are. 🤔

I'm scheduling this for the release after the planned 1.15.0 release, as I don't want to delay the 1.15.0 release even further.

/cc @tvrprasad What do you think.

chdh commented:
@arthurschreiber, thanks for you comment. I will try to simplify the PR.

chdh commented:
I have simplified the PR to apply backpressure (pause/resume) only to the last stream (the token stream parser transform) and bridge the backpressure gap between the token stream parser transform and the packet stream transform. That way we have a "minimal invasive" patch with the data flow control concept we want in the long term.

ElfenLiedGH commented:
@chdh pause's great, but what do you think about timeout? should it pause too?
I take several record, then paused and then timeout, but query is Ok. 
I'm not sure that's right

arthurschreiber commented:
From the "Client Request Timer" section of the TDS specification:

> Controls the maximum time spent waiting for a query response from the server for a client request sent after the connection has been established.

This means that as soon as we received the first byte from the response message for a query, we need to stop the request timer. If the user wants some sort of time out on the processing of the data, it's up to them to define a more suitable timer.

chdh commented:
> as soon as we received the first packet from the response message for a query, we need to stop the request timer

You are right, but according to the documentation of [SqlCommand.CommandTimeout](https://msdn.microsoft.com/en-us/library/system.data.sqlclient.sqlcommand.commandtimeout.aspx) in the .NET API, the timeout can also occur between two consecutive rows.

> This property is the cumulative time-out (for all network packets that are read during the invocation of a method) for all network reads during command execution or processing of the results. A time-out can still occur after the first row is returned, and does not include user processing time, only network read time.
For example, with a 30 second time out, if Read requires two network packets, then it has 30 seconds to read both network packets. If you call Read again, it will have another 30 seconds to read any data that it requires.

I don't know how we could implement this correctly.

arthurschreiber commented:
> You are right, but according to the documentation of SqlCommand.CommandTimeout in the .NET API, the timeout can also occur between two consecutive rows.

Weird. This is not how `mssql-jdbc` is implemented (it also follows the spec). I'd suggest we stick to the side of the specification.

chdh commented:
@ElfenLiedGH please have a look at the new version. Does it solve your problem?

tvrprasad commented:
@chdh @arthurschreiber None of the Microsoft implemented drivers implement the notion of pause/resume. The reasoning as I understand is that you can't really slow down how fast the server sends data as there is no support this throttling in TDS specification. So the memory pressure is going to build one place or the other in the system. Is there something different about NodeJS that it makes sense here? Curious to know.

@arthurschreiber Agree with not delaying release of 1.15. I'll need to read up on streaming in NodeJS before I can share thoughts on it :-)

tvrprasad commented on src/connection.js:
```diff
@@ -522,7 +534,15 @@ class Connection extends EventEmitter {
     this.messageIo = new MessageIO(this.socket, this.config.options.packetSize, this.debug);
-    this.messageIo.on('data', (data) => { this.dispatchEvent('data', data); });
+    this.messageIo.on('data', (data) => {
+      this.clearRequestTimer();
```
Can you do a separate PR for this bug fix? Though this got caught as part of this work, I think we can call it unrelated.

chdh replied:
Yes, I could put this into a separate PR, but it's not really a problem as long as you don't have a large result set or pause/resume. There was a real bug with clearRequestTimer() which I already fixed in PR #527.
@arthurschreiber what do you think?

chdh replied:
OK, it's in #530 now.

tvrprasad commented on src/request.js:
```diff
@@ -136,4 +136,26 @@ module.exports = class Request extends EventEmitter {
+  isActive() {
+    return this.connection && this.connection.request === this && this.connection.state === this.connection.STATE.SENT_CLIENT_REQUEST;
+  }
 };
```
This introduces pretty deep coupling between the Connection class and Request class. Perhaps this suggests pause/resume methods should be on the Connection class?

chdh replied:
If we have multiple active result sets (MARS) in the future, we need pause/resume per request. But you are right, the deep coupling is not clean. I will move this logic into the Connection class, but keep the pause/resume methods in the Request class.

It's a general problem that we cannot indicate public/protected/private with class methods and properties. Therefore is suggest to use TypeScript (see #512).

chdh commented:
@tvrprasad The need for pause/resume only arises in an asynchronous environment like Node.js. In the traditional programming environments with synchronous read-next calls, the data flow control is implicit. All Microsoft database drivers implement this implicit data flow control.

Example with MS-Access / VBA:

```
Dim rs As Recordset
Set rs = db.OpenRecordset("SELECT * FROM HugeTable", dbOpenForwardOnly)
Do While Not rs.EOF
  Debug.Print rs.Fields(0)
  ... wait a second ...
  rs.MoveNext
  Loop
```

In this example with MS-Access, the memory is not filled up, even if HugeTable has tens of millions of records. If you do the same thing with the current Tedious release, the memory fills up and after a few seconds the Node.js runtime will crash. Without pause/resume there is no way to control the data flow. The only solution is using SQL queries that don't generate a lot of data.

tvrprasad commented:
> In this example with MS-Access, the memory is not filled up, even if HugeTable has tens of millions of records.

Isn't the memory being filled up at the network layer (perhaps the TCP buffers), in that case? I assume the server is still sending data as fast as it can.

arthurschreiber commented:
TCP has built in flow control and the sender will slow down sending data to
the receiver automatically.

tvrprasad commented:
> TCP has built in flow control and the sender will slow down sending data to
the receiver automatically.

Right, thought about that but it was not clear to me how that'd kick in here. The TCP layer on the client would have to stop sending ACKs while it's in the pause mode to throttle the server sending data. But then I thought that would cause TCP timeouts, retransmissions and ultimately errors on the server side. Now I see that the pause has to be really long for that to happen. Presumably server has mechanisms to handle slow clients. Makes sense now. Thanks :-)

Microsoft drivers like ADO.net support Async mode API but don't have pause/resume. So they'd be running into this issue, right? I wonder why there is not a demand for that support in those drivers.

chdh commented:
> Microsoft drivers like ADO.net support Async mode API but don't have pause/resume. So they'd be running into this issue, right? I wonder why there is not a demand for that support in those drivers.

It's another concept. In the .NET API, you have to call [DbDataReader.ReadAsync()](https://msdn.microsoft.com/en-us/library/hh223681.aspx) for each row. When you stop calling ReadAsync(), the row stream is paused implicitly. With Tedious and other Node libraries, the rows are continuously delivered via events and you have to call pause() explicitly.

chdh commented on test/integration/pause-resume-test.js:
```diff
@@ -0,0 +1,239 @@
+  let paused = false;
+  openConnection();
+
+  function openConnection() {
```
OK, I have refactored the pause/resume test cases with setUp/tearDown.

tvrprasad commented:
@chdh Thanks for the clarification! Certainly learnt some stuff from this thread :-) I'll send you a few more comments on this PR.

tvrprasad commented on test/integration/pause-resume-test.js:
```diff
@@ -0,0 +1,246 @@
+      test.ok(!socketRs.flowing,
+        'Socket is not paused.');
+    }
+    test.ok(socketRs.length >= Math.min(socketRs.highWaterMark - 512, 0x4000),
```
Please define consts for these numbers to describe what they represent.

chdh replied:
OK, but it will not make it much clearer, because these are more heuristic values, not exact science. :)

tvrprasad replied:
A brief comment would be helpful.

tvrprasad commented on test/integration/pause-resume-test.js:
```diff
@@ -0,0 +1,246 @@
+    const packetTransformRs = connection.messageIo.packetStream._readableState;
+    test.ok(!packetTransformRs.flowing,
+      'Packet transform is not paused.');
+    test.ok(packetTransformWs.length <= packetTransformWs.highWaterMark && packetTransformRs.length <= packetTransformRs.highWaterMark,
```
Line too long. Perhaps you can break them into two assertions.

chdh replied:
OK

tvrprasad replied:
Thanks.

tvrprasad commented on test/integration/pause-resume-test.js:
Formatting - move } to new line.

tvrprasad commented on test/integration/pause-resume-test.js:
line:193 validates rowCount for cancelled request but rowCount is not updated here in canceled state. Should it be? What's the expectation in terms of getting 'row' events after canceling a paused request?

tvrprasad commented on test/integration/pause-resume-test.js:
I see a pause/resume after each row. Which part of the code validates this?

tvrprasad commented on src/request.js:
```diff
@@ -136,4 +136,19 @@ module.exports = class Request extends EventEmitter {
+
+  // Temporarily suspends the flow of data from the database.
+  // No more 'row' events will be emitted until resume() is called.
+  pause() {
```
Should we remove pause/resume on Request class and leave them on Connection class only? These seem roughly equivalent of cancel() which is on Connection class only.

chdh replied:
No. From the API point of view, cancel() should be a member of the Request class. It's the request that is beeing canceled, not the connection.

tvrprasad replied:
Sounds reasonable. Opened https://github.com/tediousjs/tedious/issues/534 to add 'cancel' on Request and deprecate it on Connection.

tvrprasad commented on test/integration/pause-resume-test.js:
Change '==' to '===' and '!=' to '!==' in the file?

tvrprasad commented on test/integration/pause-resume-test.js:
nit: (EOM / packet.isLast() => (EOM / packet.isLast())

tvrprasad commented on test/integration/pause-resume-test.js:
I suppose this is implicitly tested since there is a pause after the last row...

chdh commented:
@tvrprasad 

Only the two new methods Request.pause() and Request.resume() are intended to be part of the public Tedious API. All other changes are internal. The reasons that these two methods are members of the Request class and not of the Connection class are:

1. The 'row' events are emitted from the Request class and not from the Connection class. The pause/resume methods control these 'row' events. From the application point of view, the Request class represents the data source.

2. It's true that the cancel() method is currently a member of the Connection class. But that is because it has been designed from the implementation point of view and not from the application point of view. In the upper level driver node-mssql, the [cancel()](https://github.com/patriksimek/node-mssql#cancel) method is a member of the Request class. In the .NET API, the [Cancel()](https://msdn.microsoft.com/en-us/library/system.data.sqlclient.sqlcommand.cancel.aspx) method ist a member of the SqlCommand class. It's not the connection that is beeing canceled, it's the request / command. Therefore the cancel() method should not be a member of the Connection class.

3. If we later have multiple active requests at the Tedious API level (MARS or connection pooling), it must be clear which request is affected by the pause/resume call. The application logic uses the Request object as a handle for the data source and should not be forced to know the associated Connection object for using data flow control.

It's true that with the current implementation only a single Request object can be active at a time. But thats the internal view and we always have to think both sides, the internal implementation side and the external API user / application side. The application might still have old instances of Request objects that are no longer active. When pause/resume is called accidentally on an inactive Request object, the implementation must detect this and ignore the call (or throw an exception). This is the reason why the pause/resume methods check whether the associated Request object represents the currently active request.

I hope you understand the reasons for these design principles. I would like to also write a first version of a streaming BulkLoad implementation (#523), because I need that for my customer projects and I need it soon. But I don't have time to explain and justify everything in this detail.

chdh commented on src/connection.js:
```diff
@@ -522,7 +534,15 @@ class Connection extends EventEmitter {
+    this.messageIo.on('data', (data) => {
+      this.clearRequestTimer();
+      const ret = this.dispatchEvent('data', data);
+      if (ret === false && this.state === this.STATE.SENT_CLIENT_REQUEST) {
```
OK, I will move it to SENT_CLIENT_REQUEST.data.

tvrprasad replied:
Thanks.

chdh commented on src/connection.js:
```diff
@@ -708,10 +729,28 @@ class Connection extends EventEmitter {
     return this.tokenStreamParser.addBuffer(data);
   }
 
+  pauseRequest(request) {
```
This is only an internal method that is called from Request.pause(). It has to check whether this Request object represents the currently active request. I will add a comment.

tvrprasad replied:
Sounds good.

chdh commented on src/token/token-stream-parser.js:
```diff
@@ -27,14 +27,36 @@ class Parser extends EventEmitter {
 
+  // Writes an end-of-message (EOM) marker into the parser transform input
+  // queue. StreamParser will emit a 'data' event with an 'endOfMessage'
+  // pseudo token when the EON marker has passed through the transform stream.
```
Thanks.

tvrprasad replied:
np :-)

chdh commented on test/integration/pause-resume-test.js:
OK

chdh commented on test/integration/pause-resume-test.js:
No, this test is a reaction to the problem reported by Congelli501 (second message in #518),
The problem was, that the last packet with the EOM mark has already been received and processed when there are still rows in the queue of the token parser transform. With the first test case which uses 200,000 rows and stops only once after 50'000 rows, the error could not be detected, because the end of the TDS message was still far away when stop() was called. In this second test case, all the rows are probably sent within the same and only packet. After the first 100ms delay, the 'message' event has surely been emitted by the MessageIO class, because it has already detected the reception of the last packet of the message.

chdh commented on test/integration/pause-resume-test.js:
No, the end-bracket it at the end of the paragraph. But I will remove the brackets to make it clearer.

chdh commented on test/integration/pause-resume-test.js:
OK

chdh commented on test/integration/pause-resume-test.js:
The API documentation does not specify whether rows are emitted after cancel(). The current implementation does this, but it's not specified as part of the API. It also depends on what the server does when it receives the ATTENTION message. So the test case assumes that the reception of rows after cancel() is undefined and ignores them.

tvrprasad commented on test/integration/pause-resume-test.js:
Got it. Thanks for the explanation.

tvrprasad commented on test/integration/pause-resume-test.js:
ok. Thanks.

tvrprasad commented on test/integration/pause-resume-test.js:
From TDS spec:
"The client can interrupt and cancel the current request by sending an Attention message. This is also known as out-of-band data, but any TDS packet that is currently being sent MUST be finished before sending the Attention message. After the client sends an Attention message, the client MUST read until it receives an Attention acknowledgment. "

Either way, blocking or emitting remaining rows, seems consistent with the spec as long as we emit all the rows before sending the ECANCEL and the code seems to do that correctly. As far as I'm concerned we can keep the current behavior.

I would modify the test though to validate the current implementation. I think it's best to be intentional if we decide to change the implementation in future, and make the correspond test modification.

tvrprasad commented on test/integration/pause-resume-test.js:
No idea. @arthurschreiber ?

tvrprasad commented on test/integration/pause-resume-test.js:
Thanks.

tvrprasad commented:
@chdh Thanks for the detailed design rationale, makes sense and sounds good.

[rant:begin] It's too bad that we can't really stop anyone from using the internal pauseRequest/resumeRequest on the connection object. I guess that's life with JavaScript. [rant:end]

Congelli501 commented:
You can prepend protected methods by an underscore.
This convention is widely used and static code checkers (at least eslint) can check that you only use _ methods with `this`.

tvrprasad commented:
@Congelli501 Glad I ranted :-) Opened issue https://github.com/tediousjs/tedious/issues/536 to track this. Thanks!

chdh commented on test/integration/pause-resume-test.js:
After some thoughts I think it's better to suppress further 'row' events after a stopped request has been cancelled or timed out. This is what the application expects. If the application explicitly wants to receive the remaining rows on a paused request, it can call resume() before calling cancel().

tvrprasad commented on test/integration/pause-resume-test.js:
Agreed, this is a better approach. This prevents the possibility of an application _bug_ where it chokes on 'row' events coming in after 'cancel'. Making it difficult to write bugs is good :-)

chdh commented:
(With the newest Node v7.7.2, "npm test" fails because of nodejs/node#11761. It has nothing to do with this PR.)

ElfenLiedGH commented:
@chdh 

> @ElfenLiedGH please have a look at the new version. Does it solve your problem?

i have taken lasted version and "Timeout: Request failed to complete in 15000ms" on paused request

tvrprasad commented:
> Note that with the newest Node v7.7.2, "npm test" fails because of nodejs/node#11761. 

I pushed a work-around for this, see https://github.com/tediousjs/tedious/pull/538.

> I have uploaded a rebased new version of this PR with all the discussed corrections and enhancements.

It'd be helpful for incremental review if you can make rebasing it's own commit and push corrections and enhancements on top as a separate commit.

chdh commented:
@tvrprasad 
> It'd be helpful for incremental review if you can make rebasing it's own commit and push corrections and enhancements on top as a separate commit.

OK, i will do that in the future.

tvrprasad commented on src/connection.js:
```diff
@@ -1245,6 +1286,7 @@ Connection.prototype.STATE = {
     name: 'SentClientRequest',
     exit: function() {
       this.clearRequestTimer();
+      this.tokenStreamParser.resume();
```
Request class has 'paused' state now and that'll be inconsistent with the actual state of the request at this point.
Move the 'paused' state to Connection class and have Request pause/resume methods just forward the calls to Connection. That'll remove this coupling.

chdh replied:
It's more complicated than that.

1. As long as Connection.execSql() has not yet been called, the request does not know the connection.
2. When the request is cancelled or time out, we have to resume the token parser transform, but the request can still remain in the paused state.

So the paused state of the Request objects are independent of the paused state of the token parser transform.

tvrprasad replied:
> As long as Connection.execSql() has not yet been called, the request does not know the connection.

That's true. You could call this.request.resume() to keep the paused state in a consistent manner between Request class and Connection class. If that were to be the only concern...

> So the paused state of the Request objects are independent of the paused state of the token parser transform

This boils down to Request classes decides if the Request is paused state. And Connection class decides what that actually means. Sounds like a reasonable separation to me.

chdh replied:
> This boils down to Request classes decides if the Request is paused state. And Connection class decides what that actually means. Sounds like a reasonable separation to me.

Yes exactly. Whether the request is in the paused state or not is determined by the application. This is consistent from the API point of view. Whether the token parser transform and all the other internal transforms are in the paused state, is internal to the driver and not exposed to the API.

tvrprasad replied:
Agreed.

tvrprasad commented on src/request.js:
```diff
@@ -136,4 +137,27 @@ module.exports = class Request extends EventEmitter {
+    if (this.paused) {
+      return;
+    }
+    this.paused = true;
```
Move 'paused' to Connection class, keep all the logic and state in one place.

chdh replied:
The paused state of the request objects is not the same as the paused state of the token stream transform.
See above comment for the SENT_CLIENT_REQUEST exit event.

tvrprasad replied:
Alright.

chdh commented on src/connection.js:
```diff
@@ -416,7 +416,9 @@ class Connection extends EventEmitter {
           this.request.rst.push(token.columns);
         }
-        return this.request.emit('row', token.columns);
+        if (!(this.state === this.STATE.SENT_ATTENTION && this.request.paused)) {
```
Because it would change the current behaviour and break the test case connection-test / cancelRequest.

It's not only cancel(), it's also the timeout.

The current behaviour is that all the remaining rows are emitted until the server receives the ATTENTION message and stops sending rows.

And there is the effect that the server sends the rows it has buffered to far when it receives the ATTENTION message. In an interactive tool (like SQL Server Management Studio) the user would want to see the rows that have been produced so far up to the time when the user clicks the cancel button.

chdh commented on src/connection.js:
```diff
@@ -717,10 +731,33 @@ class Connection extends EventEmitter {
+  // on an old inactive Request object.
+  pauseRequest(request) {
+    if (this.isRequestActive(request)) {
+      this.tokenStreamParser.pause();
```
It's not only a test whether the request belongs to this connection. It could also be a request that is no longer active. This can happen because of a timeout. 

When the request has already timed out, has completed normally or has been previously canceled, it's better to ignore an additional cancel() call.

chdh replied:
OK, I see your point. It would be the same kind of check as in Connection.cancel(). But it's problematic.

Consider the following sequence:

- The application calls connection.execSql() to start a query.
-  The query takes longer than expected. A request timeout occurs. The driver sends ATTENTION to the server and enters the SENT_ATTENTION state. The application has no way of knowing this.
- While the driver is in the SENT_ATTENTION state and waiting for the attention response from the server, the application calls Connection.cancel(). The current implementation of the cancel() method considers this an error, writes the debug message "_Requests can only be canceled in the SentClientRequest state, not the SentAttention state_" and returns false.

From an API point of view, I would rather be forgiving than fussy. When the application wants to pause or resume a request that has already completed, then it's OK, because there are no more rows anyway that would be affected. Also when the application wants to cancel a request that has already been canceled, or close a file that is already closed, it's common to just ignore it. The driver has to protect it's inner state and the consistency of the API and therefore ignore pause/resume calls when they are not called with the currently active Request object. But it's not the purpose of the driver to educate the application.

tvrprasad replied:
> While the driver is in the SENT_ATTENTION state and waiting for the attention response from the server, the application calls Connection.cancel(). The current implementation of the cancel() method considers this an error, writes the debug message "Requests can only be canceled in the SentClientRequest state, not the SentAttention state" and returns false.

I think that's a bug and we should fix that. Opened https://github.com/tediousjs/tedious/issues/545. 

My comment is about after the Request is done and the callback is invoked. At this point the application knows the request is done. I'm not able to come up with an application pattern where there could reasonably be more calls on the Request object other than a bug in the code. Being forgiving here I think hurts than helps the developer by hiding bugs. And there is no need for a isCompleted either, as the callback already gives that information.

chdh commented on src/connection.js:
```diff
@@ -986,6 +1023,7 @@ class Connection extends EventEmitter {
       }
 
       this.request = request;
+      this.request.connection = this;
```
see my comment in #536.

And it wouldn't be enough to set only a reference to the function, because you also need the context, e.g.:

`this.request.pauseCallback = () => {this.pauseRequest(); };`

And then the question arises, whether Request.pauseCallback is a public member of the Request class. You have more or less the same ugliness and more confusion.

chdh replied:
I don't understand. Request.pauseCallback/resumeCallback would be private but still can be accessed from within the Connection class. Why can't we make Request.connection private and access it from within the Connection class and make Connection.pauseRequest()/resumeRequest() private and access it from within the Request class?

tvrprasad replied:
The difference, at least as I see it, is the coupling.

In one case Request has knowledge of Connection class and internal method names inside the Connection class. For example if I rename the methods in the Connection class or move them to a sub-class I'd have to change the Request class.

In the other case, the Connection class is just passing the callbacks to Request and Request class honors the contract by invoking them, just as it invokes user callback on completion.

chdh replied:
> The difference, at least as I see it, is the coupling.

Yes, that's true. But I would like to finish my work on this feature now and work on other things. Can't we agree that the current implementation is good enough? There is always something that can be improved and de-coupling like this has not been a design principle in Tedious until now.

tvrprasad replied:
@chdh The implementation is certainly good enough and I'll approve. The couple of tweaks left I can make them after merge. I'll let @arthurschreiber merge the change as he wanted to review before merge.

Thanks for your contribution and thanks for you patience!

chdh commented on src/connection.js:
```diff
@@ -995,7 +1033,10 @@ class Connection extends EventEmitter {
       });
-      return this.transitionTo(this.STATE.SENT_CLIENT_REQUEST);
+      this.transitionTo(this.STATE.SENT_CLIENT_REQUEST);
+      if (request.paused) {                                // Request.pause() has been called before the request was started
```
OK

arthurschreiber commented:
So just as a heads up - I'll be on vacation for a week. I'd like to give this a thorough review before this gets merged, but won't get to it until after my vacation.

chdh commented:
> So just as a heads up - I'll be on vacation for a week. I'd like to give this a thorough review before this gets merged, but won't get to it until after my vacation.

OK.  What about #523, if I implement it in the same manner? Will you accept a PR as a preliminary solution?

chdh commented:
@ElfenLiedGH 

> i have taken lasted version and "Timeout: Request failed to complete in 15000ms" on paused request

I have extended the test cases and cannot reproduce this problem. Did you include #530? Could it be that your query takes longer than 15 seconds before the first data package is sent from the server to the client? Could you provide a test case to reproduce the problem?

ElfenLiedGH commented:
@chdh 

> @ElfenLiedGH
> 
> i have taken lasted version and "Timeout: Request failed to complete in 15000ms" on paused request
> I have extended the test cases and cannot reproduce this problem. Did you include #530? Could it be that your query takes longer than 15 seconds before the first data package is sent from the server to the client? Could you provide a test case to reproduce the problem?

Oops. i took this only, now it works, thz

arthurschreiber commented:
As mentioned in https://github.com/tediousjs/tedious/pull/547, I'm trying to release `2.0.0` next week.

This would also be a great feature to include soon after! 👍

tvrprasad commented:
@arthurschreiber Now that we got the 2.0.0 released, I'll go ahead and merge this unless you have more comments.

tangkhaiphuong commented:
@arthurschreiber  Please merge this pull request and release this feature. I wait a long time this feature for resolve back-pressure. Thanks!!!

arthurschreiber commented:
@tvrprasad @tangkhaiphuong @chdh Just a quick update here: I'd _really_ like to have #528 merged before this and have this based on top of it. I believe the two changes fit nicely together and will also pave the way for other streaming changes like #523 or the new token stream reader and writer I've been working on over in https://github.com/tediousjs/node-tds-token-stream.

I'll try to get #528 in shape by the end of the week! 👍

ilusharulkov commented:
Hi! 
Sorry for my impatience, but I realy need this for sync large amount of data. Is there some forecast
when this feauture will be merged into master?

ilusharulkov commented:
Guys!
Any news ?

tangkhaiphuong commented:
@arthurschreiber  It's over 3 months but this feature still not merge, so what problem?

arthurschreiber commented:
Heya @chdh,

thank you so much for this contribution! I cleaned up the test cases a bit and fixed one tiny issue with paused requests and connection closing, but I'm really happy to have these changes be finally part of `tedious`.

I'll review the other contributions that were made since the last release and push a new version including these changes as soon as possible. I'll also try to get streaming bulk data loads into another release shortly after.

Again, thank you so much for this contribution and sorry that it has taken so long to get this merged.

❤️

chdh commented:
@arthurschreiber
Thanks for your cleanup, extended testing and improvements. I'm glad that this PR will make it into the next Tedious release.

jschell12 commented:
Hey @arthurschreiber. Any word on when you'll release to npm?

v-suhame commented:
@chdh `pause-resume-test` fails occasionally in CI, I have restarted the travis build couple of times on contributor's PR, forgot to copy the error stack though. I will share the message, if I notice it again for you to fix.

v-suhame commented:
Here is the stack
```
AssertionError: false == true
    at Object.ok (/home/travis/build/tediousjs/tedious/node_modules/nodeunit/lib/types.js:83:39)
    at Timeout._onTimeout (/home/travis/build/tediousjs/tedious/test/integration/pause-resume-test.js:114:14)
    at ontimeout (timers.js:386:14)
    at tryOnTimeout (timers.js:250:5)
    at Timer.listOnTimeout (timers.js:214:5)
```

arthurschreiber commented:
@v-suhame this is actually my fault. 😥 I’ll open a pull request to fix the flaky tests soon.

chdh commented:
@arthurschreiber @v-suhame I had similar effects with my version of the tests. I suggest to repeat the isPaused() test for a couple of times with a delay for each retry.

hosseinGanjyar commented:
Is there a way for my problem?
https://github.com/patriksimek/node-mssql/issues/471
thanks a lot

chdh commented:
@hosseinGanjyar If pause/resume was implemented in node-mssql, you could call `Request.pause()` to stop the "row" events and `Request.resume()` to resume them. I think it would be easy to implement. Node-mssql could just forward the pause/resume calls to the Tedious driver. Without that, I can't see a way to call the pause/resume methods of the underlying Tedious driver.

gsamal commented:
@chdh I am using the upper layer library `node-mssql` and they have added the pause/resume on the tedious driver. It all seems to be working until I noticed that even if request is paused, it keeps fetching rows and adds it in memory. 
An easy way to reproduce this is after n rows, hit pause, then do nothing. The memory consumption keeps growing. Does this pause/resume work on the socket/db-connection level? or how does it work? Any ides why memory keeps rows as though it looks the record are still stacking up in memory even if it's paused.

chdh commented:
@gsamal The pause mechanism works by applying back-pressure on the internal chain of streams. Each stream continues to buffer data until the high water mark is reached. Then it applies back-pressure to the next preceding stream in the chain.

Could you open a new issue and supply some more detailed information about the effect you are investigating? How much does the memory grow after pause() was called. The best way would be to provide an isolated test case that allows us to reproduce the problem,

gsamal commented:
@chdh Okay. But it seems to be keep adding records in a buffer somewhere, only pause the internal chain of streams. I can see a similar issue reported here - [node-mssql#832](https://github.com/tediousjs/node-mssql/issues/832). Can you please take a look at it? 

In a simpler sense, if I pause the request after getting n records, then I monitor the network and memory, it keeps increasing.

arthurschreiber commented:
@gsamal This was fixed via #878.

arthurschreiber commented:
The fix for this is part of the `6.0.1` release. 🎉
